### PR TITLE
Inherit XACML and use COLLECTION_POLICY for namespace

### PR DESCRIPTION
**Pull request by whikloj**
Address [ISLANDORA-1349](https://jira.duraspace.org/browse/ISLANDORA-1349)
- Inherit XACML policies from parent if possible
- Namespace is assigned by using the following (in order of precedence)
  1. namespace argument (if given)
  2. if the parent has a COLLECTION_POLICY, use that for namespaces
  3. use Fedora configured default

### CODE REVIEW

File: includes/preprocessor_base.inc [hunk 1]
```diff
@@ -107,20 +107,61 @@ abstract class IslandoraBatchPreprocessor {
...
 
     // TODO: Implement some form of caching, so we can get multiple at a time.
     return $this->connection->repository->api->m->getNextPid($namespace);
   }
 
+
+  /**
+   * Try to get the namespace from the collection policy.
+   *
+   * @param IslandoraBatchObject $object
```
Comment (adam-vessey) [inline includes/preprocessor_base.inc]: The type should match the hint in the method signature.

File: includes/islandora_scan_batch.inc [hunk 2]
```diff
@@ -520,6 +520,20 @@ EOXML;
 
     $this->relationships->add($uri, $predicate, $collection);
     $this->addContentModelRelationships();
+    $this->inheritXacmlPolicies();
+  }
+
+  /**
+   * Add any XACML policies from the parents.
```
Comment (adam-vessey) [inline includes/islandora_scan_batch.inc]: Comment seems misleading... We don't add the `POLICY` directly, we just write the relationship so it should happen later... Not sure what an ideal comment would look like...

File: includes/preprocessor_base.inc [hunk 3]
```diff
@@ -107,20 +107,61 @@ abstract class IslandoraBatchPreprocessor {
   /**
    * Get an identifier we can use for an object.
    *
+   * @param IslandoraBatchObject $object
```
Comment (adam-vessey) [inline includes/preprocessor_base.inc]: Type mismatch with hint.

File: includes/islandora_scan_batch.inc [hunk 4]
```diff
@@ -528,4 +542,12 @@ EOXML;
   protected function addContentModelRelationships() {
     $this->models = $this->preprocessorParameters['content_models'];
   }
+
+  /**
+   * Accessor for content models.
+   */
+  public function getContentModels() {
```
Comment (adam-vessey) [inline includes/islandora_scan_batch.inc]: Is this really necessary? `$this->models` should already be available anywhere it needs to be gotten.

Comment (whikloj) [inline includes/islandora_scan_batch.inc]: Fair point, or should $models be a protected variable. I was trying to mimic the use in Islandora_importer so perhaps IslandoraScanBatchObject should have something like this?
https://github.com/Islandora/islandora_importer/blob/7.x/includes/importer.inc#L224-L228

File: includes/islandora_scan_batch.inc (hunk 2 above)
Comment (whikloj) [inline includes/islandora_scan_batch.inc]: How about
`Add inheritXacmlFrom relationship.`

File: includes/preprocessor_base.inc (hunk 1 above)
Comment (whikloj) [inline includes/preprocessor_base.inc]: Sorry I saw the same thing at line 177, should the doc be updated or could we change the hint in the method?

File: includes/preprocessor_base.inc [hunk 5]
```diff
@@ -107,20 +107,61 @@ abstract class IslandoraBatchPreprocessor {
...
+   *   The object to be later processed and added to Fedora.
+   *
+   * @return string|NULL
+   *   The namespace with which the object should be ingested.
+   *   NULL if we use the Fedora configured default.
+   */
+  protected function getNamespace(AbstractFedoraObject $object) {
+    // Get namespace based on collection.
+    $collection_policies =& drupal_static(__METHOD__, array());
+    $parent_pid = $this->parameters['parent'];
```
Comment (adam-vessey) [inline includes/preprocessor_base.inc:147]: May be paranoia, but `parent` currently isn't required to be available... Really, we don't/try not to make any assumptions about what is going to be available in `$this->parameters`: https://github.com/Islandora/islandora_batch/pull/57/files#diff-728bc594d18c04ff0f4d44f6d03017bdR40

Comment (adam-vessey) [inline includes/preprocessor_base.inc:147]: Should be good if we can degrade properly to the old behaviour when we don't have it (without warnings being thrown).

Comment (whikloj) [inline includes/preprocessor_base.inc:147]: Sorry I was coming out of the addRelationships function which seemed to be assuming `parent` was there, so I didn't think too much of it. 
https://github.com/uml-digitalinitiatives/islandora_batch/blob/ISLANDORA-1349/includes/islandora_scan_batch.inc#L517

Is there another/a better way to get a batch job's parent?

Comment (adam-vessey) [inline includes/preprocessor_base.inc:147]: Really, should be fine if we just jump out before we get here if we don't have `parent`... Something like:

``` php
if (!isset($this->parameters['parent'])) {
  // Could explicitly return NULL, if desired?
  return;
}
$parent_pid = $this->parameters['parent'];
[...]
```

If we /really/ wanted to, we could make the implementation here deal with only `$this->parameters['namespace']` as it does now, and have the implementation looking at the collection policy in the scan batch preprocessor (which would make a call to `parent::getNamespace($object)` and use it if not `NULL` kind of stuff...

Comment (DiegoPino) [inline includes/preprocessor_base.inc:147]: Quick note @adam-vessey: $this->parameters['parent'] defaults to the root collection anyway
https://github.com/Islandora/islandora_batch/blob/36d41bb9436ab3670528e141e990abe4c5f259b5/islandora_batch.drush.inc#L118 ?

Comment (adam-vessey) [inline includes/preprocessor_base.inc:147]: @DiegoPino : [drush invokes a particular preprocessor implementation](https://github.com/Islandora/islandora_batch/blob/36d41bb9436ab3670528e141e990abe4c5f259b5/islandora_batch.drush.inc#L118), while the `getNamespaces()` method has been introduced into the abstract base class... I'm worried about breaking other subclasses of the abstract base which do _not_ inherit from `IslandoraScanBatch`.

**Comment by whikloj** [comment]
@adam-vessey, I hope I am correct. But I took your explanation to be that the Xacml policy stuff is fine because it is implemented via the drush call. So I only added the quick exit for the getNamespace stuff.
I hope you can have a look and maybe @ruebot will let us get it in the release. Cheers.

**Comment by adam-vessey** [comment]
Hmm... Just noticed: `islandora_importer` now throws a strict warning because it already had a [`getNamespace()` method](https://github.com/Islandora/islandora_importer/blob/7.x/includes/importer.inc#L143-L174), but with a slightly different signature.

**Comment by bradspry** [comment]
I don't know if its helpful, but it could be related; ISLANDORA-1424, Zip File Importer does not respect XACML POLICY:
https://jira.duraspace.org/browse/ISLANDORA-1424

**Comment by whikloj** [comment]
You want I should change the name? Or.....

**Comment by adam-vessey** [comment]
@bradspry: Similar issue, but not really a symptom or side-effect of anything here, since the XACML inheritance was only introduced for the "scan batch" (CLI/drush ZIP/directory batch ingest dealio); though, the same bit of code could probably be adapted and shoved into importer's preprocessor.

@whikloj: I'm not 100% sure... Removing the type hint would suffice to get rid of the warning, though it would confuse things if people go and extend the importer stuff, being not quite sure what they might be passed. Renaming here would probably be safer... `getNamespaceFromParent()` or similar, maybe?

**Comment by whikloj** [comment]
@adam-vessey sounds like a plan, I'll get a PR ready. Thanks.

**Comment by whikloj** [comment]
@adam-vessey, one more question. Is the reason for the error that islandora_importer extends the preprocessor_base? Could we modify what islandora_importer is doing to make use of this getNamespace and remove the one from islandora_importer?? 

I haven't looked but just wondering if this even a possiblity.

**Comment by DiegoPino** [comment]
@whikloj, we are extending the importer stuff, so renaming here looks like the simplest idea to avoid problems. I don't know what would happen if you modifying islandora_importer, i had to test your implementation first.

**Comment by whikloj** [comment]
@DiegoPino we extend batch as well, in fact it appears that the problem is `IslandoraBatchImporter` extending `IslandoraBatchPreprocessor`  and both use almost the same code (both called `getNamespace`) but for different object types (`IslandoraImportObject` and `AbstractFedoraObject`).

It's fine, this (batch) is a newer implementation of getNamespace so better to change it than the older more established. Just thought it might be better to have the code only once.

**Comment by DiegoPino** [comment]
@whikloj, you are right. Thanks.

**Pull request merged**